apps/system/js/nfc_manager.js

```diff
@@ -358,21 +277,31 @@
           if (activeApp.isTransitioning() || activeApp.isSheetTransitioning()) {
             return;
           }
-          // Top visible application's manifest Url is registered;
-          // Start Shrink / P2P UI and wait for user to accept P2P event
-          window.dispatchEvent(new CustomEvent('shrinking-start'));
 
-          // Setup listener for user response on P2P UI now
-          window.addEventListener('shrinking-sent', this);
+          this._initP2PUI();
         } else {
-          // Clean up P2P UI events
-          this._logVisibly('CheckP2PRegistration failed');
-          window.removeEventListener('shrinking-sent', this);
-          window.dispatchEvent(new CustomEvent('shrinking-stop'));
+          this.debug('CheckP2PRegistration failed');
+          this._cleanP2PUI();
         }
       });
     },
 
+    _cleanP2PUI: function() {
+      window.removeEventListener('shrinking-sent', this._handleShrinkingSent);
+      window.dispatchEvent(new CustomEvent('shrinking-stop'));
+    },
+
+    _initP2PUI: function() {
+      window.dispatchEvent(new CustomEvent('shrinking-start'));
+
+      this._handleShrinkingSent = () => {
+        this._cleanP2PUI();
+        this._dispatchP2PUserResponse();
+      };
+
+      window.addEventListener('shrinking-sent', this._handleShrinkingSent);
```

Comment:
I forgot why we need to dynamically add/remove event listener here. Is there any chance we could always listen to it?